### PR TITLE
[15.0][FIX] delivery_package_number: Don't reset number on backorder

### LINKED ISSUE #676: delivery_package_number bug module when creating a backorder
## Module
delivery_package_number

## Describe the bug
We have detected a bug in the delivery_package_number module when creating a backorder.

## To Reproduce
**Affected versions**:

1. Create a sales order containing at least one storable product and quantity greater than 1, example 10.
2. Confirm the sales order.
3. Go to the created delivery.
4. Enter the 'Number of Packages' in the 'Additional Info' tab, example 2.
5. Enter the packages in the 'Done' column less than 'Reserved', example 2.
6. Click on Validate and click on 'Generate Backorder' or 'No backorder' .

**Expected behavior**
If we go again to the tab 'Additional Info' the number of packages is 1 (incorrect) instead of 2 (correct).

**Additional context**
While trying to solve it we have detected that the problem is caused by the repeated execution of the _compute_number_of_packages function that recalculates the number of packages based on package_ids and sets the Number of Packages to 1 if there are no package_ids.

## Proposal
We propose the following changes to the delivery_package_number module to fix the bug:

- Create a boolean field that will appear next to the 'Number of Packages' to be able to switch between inserting the number manually or having it automatically calculated based on the package_ids. This way we avoid recalculating the 'Number of Packages' if it is in manual mode.
- On the other hand, if the company has disabled the 'Packages' in the Inventory settings, by default the option will be manual.
- In addition, we will also add a Boolean field in the 'Operation Type' to set whether it is calculated automatically or manually by default depending on the 'Type of Operation'. In this way, we will always be able to choose from the picking whether the 'Number of Packages' is inserted manually or automatically but, depending on the 'Type of Operation', there will be a default option.

### PR DESCRIPTION
**Pull request by pedrobaeza**
Fixes #676

@Tecnativa

### CODE REVIEW

**Review by chienandalu — APPROVED** [review]

File: delivery_package_number/models/stock_picking.py [hunk 1]
```diff
@@ -11,9 +12,12 @@ class StockPicking(models.Model):
         compute="_compute_number_of_packages",
         readonly=False,
         store=True,
+        default=1,
```
Comment (rousseldenis) [inline delivery_package_number/models/stock_picking.py]: @pedrobaeza Would you want to say there is always a package (which is not the case...)?

**Comment by ValentinVinagre** [comment]
Functional Review.
When a partial delivery note is created, it does not ask for the number of packages. This case is what the issue refers to, if I'm not mistaken.

Selección_045
image


On the other hand, if you set the number of packages to 1 and you have a carrier, the user will never be asked if it is validated without entering the quantities manually (I think the number of packages would have to be 0 by default, the same as before).

**Comment by pedrobaeza** [comment]
OK, having the prompt on the dialog to enter the number of packages seems enough. Put default=0.

**Review by HaraldPanten — APPROVED** [review]
LGTM tested in runboat 👍

**Review by HaraldPanten — CHANGES_REQUESTED** [review]
Wait, it's not working. Try:

1- Create a sale order with 2 lines. 2 storable products with quantity 10 (for example), each one.
2- Add shipping.
3- Validate the SO.
4- Set 2 packages in the additional info field.
5- Set quantity 2 to the delivered quantity, only for 1 of the products.
6- Create backorder, or don't create backorder.

Result: The number of packages that we set before is changed to 1. 🤔

Tested in runboat.

**Comment by pedrobaeza** [comment]
I have tried in local and it's working fine, and I have also made a test case in local for your steps and they are working, so one of these 2 is happening:

- The runboat that this is running doesn't contain the latest code.
- There's another module that is colliding with this one. This second seems improbable, as the CIs with the tests are green.

**Review by HaraldPanten — APPROVED** [review]
My colleagues told me the same. In a local environment seems to be working fine.

@angelgarciadelachica WDYT?

**Comment by pedrobaeza** [comment]
Let's merge it then:

/ocabot merge patch

**Comment by angelgarciadelachica** [comment]
In local it works and the tests are passed so it is probably what pedro says.
LGTM

**Comment by pedrobaeza** [comment]
@sbidoul is https://github.com/OCA/delivery-carrier/pull/696#issuecomment-1719968270 something to worry about?

**Comment by sbidoul** [comment]
@pedrobaeza sorry I don't really understand that comment. Can you elaborate?

**Comment by pedrobaeza** [comment]
The problem is that the runboat linked to this PR didn't contain the patch we were applying. It's like the image taken was the one from the main branch.

**Comment by sbidoul** [comment]
> The problem is that the runboat linked to this PR didn't contain the patch we were applying. It's like the image taken was the one from the main branch.

By patch you mean this PR ? I don't see how that could happen.

**Comment by HaraldPanten** [comment]
Seems that when we were testing the feature in runboat (https://github.com/OCA/delivery-carrier/pull/696#pullrequestreview-1621632186) It was not working as expected. But it was working fine in a local environment.

On the other hand tests were correct, so it means that this branch (using this PR feature) was suposed to be working correctly as well (but it didn't in runboat).

Due to that, Pedro was wondering if some of these two options might be occuring: https://github.com/OCA/delivery-carrier/pull/696#issuecomment-1719968270

Taking in consideration that tests were OK, our doubts are about if runboat could be running without the latest code. Is that possible?

THX @sbidoul

**Comment by sbidoul** [comment]
> doubts are about if runboat could be running without the latest code. Is that possible?

I very much doubt runboat would start a build with the wrong code. The runboat part that fetches the code is here https://github.com/sbidoul/runboat/blob/90513736535b20df9afba4560d5e345eba973c3b/src/runboat/kubefiles/runboat-clone-and-install.sh#L16C60-L16C75, where RUNBOAT_GIT_REF is the commit sha.

The only thing that I could imagine are the the status updates would be delayed and you tested with a previous version of the PR?

**Comment by pedrobaeza** [comment]
> The only thing that I could imagine are the the status updates would be delayed and you tested with a previous version of the PR?

Yes, this is possible, but the link in GH status was only showing one build (not live now: https://runboat.odoo-community.org/builds/b97ff1cbb-2d82-49fa-8f06-ae5aff85a820?live)

**Comment by HaraldPanten** [comment]
Same problem is happening here (https://github.com/OCA/delivery-carrier/pull/701), if you need another example.

THX

**Pull request merged**